pyface/ui/wx/layout_widget.py

```diff
@@ -0,0 +1,121 @@
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+import typing as t
+
+import wx
+
+from traits.api import provides
+
+from pyface.i_layout_item import DEFAULT_SIZE
+from pyface.i_layout_widget import ILayoutWidget, MLayoutWidget
+from pyface.ui.wx.widget import Widget
+
+
+#: The special "default" size for Wx Size objects.
+WX_DEFAULT_SIZE = -1
+
+
+@provides(ILayoutWidget)
+class LayoutWidget(MLayoutWidget, Widget):
+    """ A widget which can participate as part of a layout.
+
+    This is an abstract class, as Widget._create_control needs to be
+    implemented at a minimum.
+    """
+
+    def _set_control_minimum_size(self, size):
+        """ Set the minimum size of the control. """
+        wx_size = _size_to_wx_size(size)
+        self.control.SetMinSize(wx_size)
+
+    def _get_control_minimum_size(self):
+        """ Get the minimum size of the control. """
+        wx_size = self.control.GetMinSize()
+        return _wx_size_to_size(wx_size)
+
+    def _set_control_maximum_size(self, size):
+        """ Set the maximum size of the control. """
+        wx_size = _size_to_wx_size(size)
+        self.control.SetMaxSize(wx_size)
+
+    def _get_control_maximum_size(self):
+        """ Get the maximum size of the control. """
+        wx_size = self.control.GetMaxSize()
+        return _wx_size_to_size(wx_size)
+
+    def _set_control_stretch(self, size_policy):
+        """ Set the stretch factor of the control.
+
+        In Wx the stretch factor is set at layout time and can't be changed
+        without removing and adding the widget back into a layout.
+        """
+        pass
+
+    def _get_control_stretch(self):
+        """ Get the stretch factor of the control.
+
+        In Wx the stretch factor is set at layout time and can't be obtained
+        from the widget itself.  As a result, this method simply returns the
+        current trait value.  This method is only used for testing.
+        """
+        return self.stretch
+
+    def _set_control_size_policy(self, size_policy):
+        """ Set the size policy of the control
+
+        In Wx the size policy is set at layout time and can't be changed
+        without removing and adding the widget back into a layout.
+        """
+        pass
+
+    def _get_control_size_policy(self):
+        """ Get the size policy of the control
+
+        In Wx the size policy is set at layout time and can't be obtained from
+        the widget itself.  As a result, this method simply returns the
+        current trait value.  This method is only used for testing.
+        """
+        return self.size_policy
+
+
+def _size_to_wx_size(size: t.Tuple[int, int]) -> wx.Size:
```

Comment:
Any specific reason why you made use of `typing` here? We haven't used it in other parts of the codebase or in other ETS libraries.

Reply:
It was an experiment on my part to see if it adds anything to use explicitly typed code and to play with patterns for doing so. This seemed like a fairly harmless corner to try it in, and one where the constraints are fairly clear.

Typing seems to be catching on, so I think we may well end up with our codebases being typed in the 2-3 year timeframe, so experiments are valuable.

If you feel strongly it should go, I will remove.

Reply:
I've removed it.  At some point I may try with a more self-contained library.